gk/main.c

```diff
@@ -16,22 +16,403 @@
+gk_process_request(struct flow_entry *fe,
+	 __attribute__((unused)) struct ipacket *packet)
+{
+	uint64_t now = rte_rdtsc();
+	uint8_t priority = priority_from_delta_time(now,
+			fe->u.request.last_packet_seen_at);
+
+	fe->u.request.last_packet_seen_at = now;
+
+	/*
+	 * The reason for using "<" instead of "<=" is that the equal case 
+	 * means that the source has waited enough time to have the same 
+	 * last priority, so it should be awarded with the allowance.
+	 */
+	if (priority < fe->u.request.last_priority &&
+			fe->u.request.allowance > 0) {
+		fe->u.request.allowance--;
+		priority = fe->u.request.last_priority;
+	} else {
+		fe->u.request.last_priority = priority;
+		fe->u.request.allowance = START_ALLOWANCE - 1;
+	}
+
+	/*
+	 * Adjust @priority for the DSCP field.
+	 * DSCP 0 for legacy packets; 1 for granted packets; 
+	 * 2 for capability renew; 3-63 for requests.
+	 */
+	priority += 3;
+	if (unlikely(priority > 63))
+		priority = 63;
+
+	/* The assigned priority is @priority. */
+	
+	/* TODO Encapsulate the packet as a request. */
+
+	/* TODO Put this encapsulated packet in the request queue. */
+
+	return 0;
+}
+
+static inline uint64_t
+cycle_from_second(uint64_t time)
+{
+	return (cycles_per_sec * time);
+}
+
+static int
+gk_process_granted(struct flow_entry *fe, struct ipacket *packet)
+{
+	bool renew_cap;
+	uint64_t now = rte_rdtsc();
+	struct rte_mbuf *pkt = packet->pkt;
+
+	if (now >= fe->u.granted.cap_expire_at) {
+		reinitialize_flow_entry(fe, now);
+		return gk_process_request(fe, packet);
+	}
+
+	if (now >= fe->u.granted.budget_renew_at) {
+		fe->u.granted.budget_renew_at = now + cycle_from_second(1);
+		fe->u.granted.budget_byte = fe->u.granted.tx_rate_kb_cycle * 1024;
+	}
+
+	if (pkt->data_len > fe->u.granted.budget_byte)
+		return drop_packet(pkt);
+
+	fe->u.granted.budget_byte -= pkt->data_len;
+	renew_cap = now >= fe->u.granted.send_next_renewal_at;
+	if (renew_cap)
+		fe->u.granted.send_next_renewal_at = now +
+			fe->u.granted.renewal_step_cycle;
+
+	/*
+	 * TODO Encapsulate packet as a granted packet,
+	 * mark it as a capability renewal request if @renew_cap is true,
+	 * enter destination according to @fe->u.granted.grantor_id,
+	 * and put the encapsulated packet in the granted queue.
+	 */
+
+	return 0;
+}
+
+static int
+gk_process_declined(struct flow_entry *fe, struct ipacket *packet)
+{
+	uint64_t now = rte_rdtsc();
+
+	if (unlikely(now >= fe->u.declined.expire_at)) {
+		reinitialize_flow_entry(fe, now);
+		return gk_process_request(fe, packet);
+	}
+
+	return drop_packet(packet->pkt);
+}
+
+static int
+setup_gk_instance(unsigned int lcore_id, struct gk_config *gk_conf)
+{
+	int  ret;
+	char ht_name[64];
+	unsigned int block_idx = lcore_id - gk_conf->lcore_start_id;
+
+	struct gk_instance *instance = &gk_conf->instances[block_idx];
+	struct rte_hash_parameters ip_flow_hash_params = {
+		.entries = gk_conf->flow_ht_size,
+		.key_len = sizeof(struct ip_flow),
+		.hash_func = DEFAULT_HASH_FUNC,
```

Comment:
Indeed, this is a potential bug. I have changed the custom function to ``rss_hash_func()``. 

Cody, could you please change the two library functions in DPDK ``static inline uint32_t rte_softrss(uint32_t *input_tuple, uint32_t input_len, const uint8_t *rss_key)`` to ``static inline uint32_t rte_softrss(const uint32_t *input_tuple, uint32_t input_len, const uint8_t *rss_key)`` as well as ``rte_softrss_be()`` with ``const`` qualifier? The two functions are in ``lib/librte_hash/rte_thash.h`` file. It seems that I cannot push updates to the DPDK library.